crates/fuel-core/src/service/genesis/runner.rs

```diff
@@ -0,0 +1,666 @@
+use fuel_core_chain_config::Group;
+use fuel_core_storage::transactional::Transaction;
+use std::sync::Arc;
+use tokio::sync::Notify;
+use tokio_util::sync::CancellationToken;
+
+use crate::database::{
+    database_description::on_chain::OnChain,
+    genesis_progress::GenesisResource,
+    transaction::DatabaseTransaction,
+    Database,
+};
+
+pub trait TransactionOpener {
+    fn transaction(&mut self) -> DatabaseTransaction<OnChain>;
+    fn view_only(&self) -> &Database;
+}
+
+impl TransactionOpener for Database {
+    fn transaction(&mut self) -> DatabaseTransaction<OnChain> {
+        Database::transaction(self)
+    }
+
+    fn view_only(&self) -> &Database {
+        self
+    }
+}
+
+pub struct GenesisRunner<Handler, Groups, TxOpener> {
+    resource: GenesisResource,
+    handler: Handler,
+    tx_opener: TxOpener,
+    skip: usize,
+    groups: Groups,
+    finished_signal: Option<Arc<Notify>>,
+    cancel_token: CancellationToken,
+}
+
+pub trait ProcessState<T> {
+    fn process(&mut self, item: T, tx: &mut Database) -> anyhow::Result<()>;
+}
+
+pub trait ProcessStateGroup<T> {
+    fn process_group(&mut self, group: Vec<T>, tx: &mut Database) -> anyhow::Result<()>;
+}
+
+impl<T, K> ProcessStateGroup<K> for T
+where
+    T: ProcessState<K>,
+{
+    fn process_group(&mut self, item: Vec<K>, tx: &mut Database) -> anyhow::Result<()> {
+        item.into_iter().try_for_each(|item| self.process(item, tx))
+    }
+}
+
+pub trait HandlesGenesisResource {
```

Comment:
Optional: I think the type that implements `ProcessState` and `HandlesGenesisResource` can be the same. It means we will create `Handler<T>` for the type `T`, and the handler will know how to process the type instead of the type itself.

Reply:
I've made the abstractions more straightforward at the cost of slight code duplication and a some more verbosity in the tests.